feat: load SharePoint Pages content, feat: load docs from root folder in drive, feat: optionally only load specific file types.

CraftingLevi opened the pull request:
# Description

Added functionality to load more data from a Sharepoint Site. 

- Added functionality to download files from the root folder of a Sharepoint Site
- Added functionality to download the HTML content of Sharepoint Pages for a Sharepoint Site
- Added functionality to limit downloading of files to specific file types.

Fixes # (issue)
#936
#937
#938

## Type of Change

Please delete options that are not relevant.

- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:


- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

Commit:
added functionality to load all documents, added functionality to load all pages

Commit:
updated with new 'include' argument and 'file_types' argument to filter file_types when loading documents.

Commit:
Added documentation

Commit:
running format and lint

CraftingLevi commented:
The tests fail in _extract_page, the output type hints states ‘None | Dict[str…’ , because sometimes a SharePoint page has no ‘TextWebParts’, meaning there is no valid text to extract from the HTML. When this is the case, None is returned.

instead, this situation should raise a ValueError that is handled by _download_pages_and_extract_metadata by passing on that page.

Commit:
fixed issue with None in output type hints

CraftingLevi commented:
Tested locally after installing test_requirements.txt, got a missing dependency error (llmsherpa), which is likely an issue with test_requirements.txt.

After installing llmsherpa, tests run with 91 passed, 13 skipped, 1 warning in 17.43s

Commit:
running format and lint

CraftingLevi commented:
I found out there is a way to do batch requests. I have an implementation running now that, for a sharepoint with +200 sites cuts down the download step from 1 minute to 7 seconds. 

Please wait with running tests until this is committed.

Commit:
implemented batch call for page content

Commit:
Improved type hints, documentation and comments

CraftingLevi committed:
Merge branch 'run-llama:main' into main

Commit:
Merge branch 'main' of https://github.com/CraftingLevi/llama-hub

CraftingLevi commented:
Successfully implemented the use of batch requests to retrieve the content of pages, significantly speeding up retrieval of page content. 

Ran all tests locally with result 91 passed, 13 skipped, 2 warnings in 8.73s, after format and lint.

Commit:
fix typo

anoopshrma commented on llama_hub/microsoft_sharepoint/README.md:
```diff
@@ -20,12 +20,18 @@ More info on Microsoft Graph APIs - [Refer here](https://learn.microsoft.com/en-
+- Load files present in a specific folder in SharePoint
+- Load all files present in the drive of a SharePoint
+- Load all pages under a SharePoint site
+
```
Line no 28 can be removed as you have already mentioned this as the first point!

anoopshrma commented on llama_hub/microsoft_sharepoint/base.py:
```diff
@@ -74,11 +82,12 @@ def _get_access_token(self) -> str:
         Retrieves the site ID of a SharePoint site using the provided site name.
 
         Args:
+            access_token (str): access_token
```
Since    `_get_site_id_with_host_name` does not have access_token as a argument. We can remove this from the docstring as well

anoopshrma commented on llama_hub/microsoft_sharepoint/base.py:
```diff
@@ -154,9 +162,12 @@ def _get_sharepoint_folder_id(self, folder_path: str) -> str:
-            f"{self._drive_id_endpoint}/{self._drive_id}/root:/{folder_path}"
-        )
+        if folder_path == "root":
+            folder_id_endpoint = f"{self._drive_id_endpoint}/{self._drive_id}/root"
```
Both if/else can be formatted in the same manner. That is either to use "(" on both or not.

CraftingLevi replied:
I adjusted the way the default argument works to not use `root`, but `""`. Now, the default endpoint is the endpoint that was under `if folder_path == "root"`, and if a `folder_path` is provided (so not ""), then the folder_path is correctly added to the default endpoint. (see latest commit to base)

anoopshrma commented on llama_hub/microsoft_sharepoint/base.py:
```diff
@@ -351,32 +520,62 @@ def get_metadata(filename: str) -> Any:
         Args:
             sharepoint_site_name (str): The name of the SharePoint site.
             sharepoint_folder_path (str): The path of the folder in the SharePoint site.
+                                          If `root`, loads data from the root folder of the
```
You should update this in the readme as well. That it'll by default extract the files from root.

CraftingLevi replied:
Updated the readme with the new logic implemented in the latest commit to base

anoopshrma reviewed (approved):
This looks great @CraftingLevi! Have added minor comments that needs to be addressed. Then we can move ahead and merge

Commit:
Updated ReadMe

Commit:
Updated docstring _get_site_id_with_host_name

CraftingLevi committed:
Merge branch 'run-llama:main' into main

Commit:
changed default sharepoint_folder_path argument to "" instead of root

Commit:
Updated ReadMe with more explicit instructions on how to use parameters, and what the default parameters are.

CraftingLevi commented:
@anoopshrma I've added all the requested changes and synced the fork prior. 

In summary: I've adjusted the default argument of 'root' for 'sharepoint_folder_path' to "", in case someone ever wants to have a folder in the root folder called 'root', then this allows for that to happen. And it works nicely with the if/else comment you've made. 

Ran make format, make lint and make test, seems all good.

Commit:
fix typo

anoopshrma commented:
Hey @CraftingLevi ,
Highly appreciate you contribution here. 
With recent llama_index update to v0.10.x, llama_hub is now deprecated. All the loaders have now moved into llamaindex repo. 

Could you push your changes there directly!! 
That would help a lot! 
https://github.com/run-llama/llama_index/tree/main/llama-index-integrations/readers/llama-index-readers-microsoft-sharepoint
